libkmod/libkmod-builtin.c

```diff
@@ -1,333 +1,185 @@
 }
 
-static off_t get_string(struct kmod_builtin_iter *iter, off_t offset,
-			char **line, size_t *size)
+static ssize_t get_string(struct kmod_builtin_info *info)
 {
-	int sv_errno;
-	char buf[BUFSIZ];
-	char *nullp = NULL;
-	size_t linesz = 0;
-
-	while (!nullp) {
-		ssize_t sz;
-		size_t partsz;
-
-		sz = pread(iter->file, buf, BUFSIZ, offset);
-		if (sz < 0) {
-			sv_errno = errno;
-			goto fail;
-		} else if (sz == 0) {
-			offset = 0;
-			break;
-		}
-
-		nullp = memchr(buf, '\0', (size_t) sz);
-		partsz = (size_t)((nullp) ? (nullp - buf) + 1 : sz);
-		offset += (off_t) partsz;
-
-		if (iter->bufsz < linesz + partsz) {
-			void *tmp;
-			size_t s;
-
-			s = linesz + partsz;
-			tmp = realloc(iter->buf, s);
+	ssize_t len;
 
-			if (!tmp) {
-				sv_errno = errno;
-				goto fail;
-			}
-			iter->bufsz = s;
-			iter->buf = tmp;
-		}
-
-		strncpy(iter->buf + linesz, buf, partsz);
-		linesz += partsz;
-	}
-
-	if (linesz) {
-		if (iter->buf[linesz - 1] != '\0') {
-			sv_errno = EINVAL;
-			goto fail;
-		}
-		*line = iter->buf;
-		*size = linesz;
+	len = getdelim(&info->buf, &info->bufsz, '\0', info->fp);
+	if (len > 0 && info->buf[len] != '\0') {
+		errno = EINVAL;
+		len = -1;
 	}
 
-	return offset;
-fail:
-	errno = sv_errno;
-	return -1;
+	return len;
 }
 
-static bool kmod_builtin_iter_next(struct kmod_builtin_iter *iter)
+static ssize_t get_strings(struct kmod_builtin_info *info, const char *modname,
+				struct strbuf *buf)
 {
-	char *line,  *modname;
-	size_t linesz;
-	off_t pos, offset, modlen;
-
-	modname = NULL;
+	const size_t modlen = strlen(modname);
+	ssize_t count = 0, n;
 
-	iter->nstrings = 0;
-	offset = pos = iter->next;
+	do {
+		char *dot, *line;
 
-	while (offset < iter->size) {
-		char *dot;
-		off_t len;
-
-		offset = get_string(iter, pos, &line, &linesz);
-		if (offset <= 0) {
-			if (offset)
-				ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-			pos = iter->size;
+		n = get_string(info);
+		if (n == -1)
 			break;
-		}
 
+		line = info->buf;
 		dot = strchr(line, '.');
-		if (!dot) {
-			ERR(iter->ctx, "kmod_builtin_iter_next: unexpected string without modname prefix\n");
-			pos = iter->size;
-			break;
+		if (dot == NULL) {
+			count = -EINVAL;
+			ERR(info->ctx, "get_strings: "
+			    "unexpected string without modname prefix\n");
+			return count;
 		}
-
-		len = dot - line;
-
-		if (!modname) {
-			modname = strdup(line);
-			modlen = len;
-		} else if (modlen != len || strncmp(modname, line, len)) {
+		if (strncmp(line, modname, modlen) ||
+		    line[modlen] != '.') {
+			if (count == 0)
+				continue;
 			break;
 		}
+		if (!strbuf_pushchars(buf, dot + 1) ||
+		    !strbuf_pushchar(buf, '\0')) {
+			count = -errno;
+			ERR(info->ctx, "get_strings: "
+			    "failed to append modinfo string\n");
+			return count;
+		}
+		if (count == INTPTR_MAX) {
+			count = -ENOMEM;
+			ERR(info->ctx, "get_strings: "
+			    "too many modinfo strings encountered\n");
+			return count;
+		}
+		count++;
+	} while (n != -1);
```

Comment:
I dislike this do-while as well because it's really just a for(;;)-loop without triggering static analyzers, I guess. I didn't think about putting count into the for-loop because it implies that count is somehow relevant for further iterations. But then again, it still looks nicer than this one and for the unlikely overflow-event, it's actually relevant. Applied.

Reply:
It was not meant that I tried to obscure an endless loop, but trying to say through code that we loop here until no more lines are left (or calling break for optimization). So I've tried to bend my mind around "how to state that we are iterating through lines here?" but the for-loop with count is still nice, since it completely avoids this obscure endless loop and implies that we would eventually find an end if there are WAY too many strings around.